frontend/src/component/project/Project/ProjectFeatureToggles/FeatureToggleSwitch/FeatureToggleSwitch.tsx

```diff
@@ -1,61 +1,223 @@
 import { UPDATE_FEATURE_ENVIRONMENT } from 'component/providers/AccessProvider/permissions';
 import { useOptimisticUpdate } from './hooks/useOptimisticUpdate';
 import { flexRow } from 'themes/themeStyles';
+import { ENVIRONMENT_STRATEGY_ERROR } from 'constants/apiErrors';
+import { formatUnknownError } from 'utils/formatUnknownError';
+import useFeatureApi from 'hooks/api/actions/useFeatureApi/useFeatureApi';
+import { useFeature } from 'hooks/api/getters/useFeature/useFeature';
+import useToast from 'hooks/useToast';
+import { useChangeRequestsEnabled } from 'hooks/useChangeRequestsEnabled';
+import { useChangeRequestToggle } from 'hooks/useChangeRequestToggle';
+import { EnableEnvironmentDialog } from 'component/feature/FeatureView/FeatureOverview/FeatureOverviewSidePanel/FeatureOverviewSidePanelEnvironmentSwitches/FeatureOverviewSidePanelEnvironmentSwitch/EnableEnvironmentDialog';
+import { UpdateEnabledMessage } from 'component/changeRequest/ChangeRequestConfirmDialog/ChangeRequestMessages/UpdateEnabledMessage';
+import { ChangeRequestDialogue } from 'component/changeRequest/ChangeRequestConfirmDialog/ChangeRequestConfirmDialog';
 
 const StyledBoxContainer = styled(Box)<{ 'data-testid': string }>(() => ({
     mx: 'auto',
     ...flexRow,
 }));
 
 interface IFeatureToggleSwitchProps {
-    featureName: string;
+    featureId: string;
     environmentName: string;
     projectId: string;
     value: boolean;
-    onToggle: (
+    showInfoBox?: () => void;
+    onToggle?: (
         projectId: string,
         feature: string,
         env: string,
         state: boolean
-    ) => Promise<void>;
+    ) => void;
 }
 
 export const FeatureToggleSwitch: VFC<IFeatureToggleSwitchProps> = ({
     projectId,
-    featureName,
+    featureId,
     environmentName,
     value,
     onToggle,
+    showInfoBox,
 }) => {
+    const { toggleFeatureEnvironmentOn, toggleFeatureEnvironmentOff } =
+        useFeatureApi();
+    const { setToastData, setToastApiError } = useToast();
+    const { isChangeRequestConfigured } = useChangeRequestsEnabled(projectId);
+    const {
+        onChangeRequestToggle,
+        onChangeRequestToggleClose,
+        onChangeRequestToggleConfirm,
+        changeRequestDialogDetails,
+    } = useChangeRequestToggle(projectId);
     const [isChecked, setIsChecked, rollbackIsChecked] =
         useOptimisticUpdate<boolean>(value);
 
-    const onClick = () => {
-        setIsChecked(!isChecked);
-        onToggle(projectId, featureName, environmentName, !isChecked).catch(
-            rollbackIsChecked
+    const [showEnabledDialog, setShowEnabledDialog] = useState(false);
+    const { feature } = useFeature(projectId, featureId);
+
+    const disabledStrategiesCount =
+        feature?.environments
+            .find(env => env.name === environmentName)
+            ?.strategies.filter(strategy => strategy.disabled).length ?? 0;
+
+    const callback = () => {
+        onToggle &&
```

Comment:
onToggle?.(projectId, feature.name, environmentName, !isChecked);

Reply:
The callback looses the semantics. All the other fns and variables are named specifically and callback is too generic. Once we introduce the refactoring I suggested above we may even get rid of the extra function